pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -1340,77 +1339,130 @@ func (n *NetworkPolicyController) getClusterGroupWorkloads(group *antreatypes.Gr
+	}
+
+	switch key.Type {
+	case controlplane.AntreaClusterNetworkPolicy:
+		cnp, err := n.cnpLister.Get(key.Name)
+		if err != nil {
+			if oldInternalNetworkPolicyObj == nil {
+				return nil
+			}
+			n.deleteInternalNetworkPolicy(oldInternalNetworkPolicy)
+			return nil
+		}
+		newInternalNetworkPolicy = n.processClusterNetworkPolicy(cnp)
+	case controlplane.AntreaNetworkPolicy:
+		anp, err := n.anpLister.NetworkPolicies(key.Namespace).Get(key.Name)
+		if err != nil {
+			if oldInternalNetworkPolicyObj == nil {
+				return nil
+			}
+			n.deleteInternalNetworkPolicy(oldInternalNetworkPolicy)
+			return nil
+		}
+		newInternalNetworkPolicy = n.processAntreaNetworkPolicy(anp)
+	case controlplane.K8sNetworkPolicy:
+		knp, err := n.networkPolicyLister.NetworkPolicies(key.Namespace).Get(key.Name)
+		if err != nil {
+			if oldInternalNetworkPolicyObj == nil {
+				return nil
+			}
+			n.deleteInternalNetworkPolicy(oldInternalNetworkPolicy)
+			return nil
+		}
+		newInternalNetworkPolicy = n.processNetworkPolicy(knp)
+	}
+
+	newAppliedToGroups := sets.NewString(newInternalNetworkPolicy.AppliedToGroups...)
+	newAddressGroups := newInternalNetworkPolicy.GetAddressGroups()
+	newNodeNames := sets.NewString()
 	// Calculate the set of Node names based on the span of the
 	// AppliedToGroups referenced by this NetworkPolicy.
-	for _, appliedToGroupName := range internalNP.AppliedToGroups {
+	for appliedToGroupName := range newAppliedToGroups {
 		appGroupObj, found, _ := n.appliedToGroupStore.Get(appliedToGroupName)
 		if !found {
 			continue
 		}
 		appGroup := appGroupObj.(*antreatypes.AppliedToGroup)
-		utilsets.MergeString(nodeNames, appGroup.SpanMeta.NodeNames)
-	}
-	updatedNetworkPolicy := &antreatypes.NetworkPolicy{
-		UID:                   internalNP.UID,
-		Name:                  internalNP.Name,
-		SourceRef:             internalNP.SourceRef,
-		Rules:                 internalNP.Rules,
-		AppliedToGroups:       internalNP.AppliedToGroups,
-		Priority:              internalNP.Priority,
-		TierPriority:          internalNP.TierPriority,
-		AppliedToPerRule:      internalNP.AppliedToPerRule,
-		PerNamespaceSelectors: internalNP.PerNamespaceSelectors,
-		SpanMeta:              antreatypes.SpanMeta{NodeNames: nodeNames},
-		Generation:            internalNP.Generation,
-	}
-	klog.V(4).Infof("Updating internal NetworkPolicy %s with %d Nodes", key, nodeNames.Len())
-	n.internalNetworkPolicyStore.Update(updatedNetworkPolicy)
-	// Internal NetworkPolicy update is complete. Safe to unlock the
-	// critical section.
-	n.internalNetworkPolicyMutex.Unlock()
-	if nodeNames.Equal(oldNodeNames) {
-		// Node span for internal NetworkPolicy was not modified. No need to enqueue
-		// AddressGroups.
-		klog.V(4).Infof("Internal NetworkPolicy %s Node span remains unchanged. No need to enqueue AddressGroups.", key)
-		return nil
+		utilsets.MergeString(newNodeNames, appGroup.NodeNames)
```

Comment:
Maybe not related to this change, just come up with this question while reviewing.
Is there a chance that when we read the `appGroup.NodeNames` here, the `NodeNames` hasn't been calculated by `func (n *NetworkPolicyController) syncAppliedToGroup(key string) error`

Reply:
NVM. Stupid question.